api/user.go

```diff
@@ -1621,33 +1622,42 @@ func resetPassword(c *Context, w http.ResponseWriter, r *http.Request) {
 	}
 
 	userId := props["user_id"]
-	hash := props["hash"]
-	timeStr := ""
 
-	if !c.IsSystemAdmin() {
-		if len(hash) == 0 {
-			c.SetInvalidParam("resetPassword", "hash")
+	if !c.IsSystemAdmin() || len(userId) == 0 {
+		code := props["code"]
+		if len(code) != model.PASSWORD_RECOVERY_CODE_SIZE {
+			c.SetInvalidParam("resetPassword", "code")
 			return
 		}
 
-		data := model.MapFromJson(strings.NewReader(props["data"]))
+		if result := <-Srv.Store.PasswordRecovery().GetByCode(code); result.Err != nil {
+			c.Err = model.NewLocAppError("resetPassword", "api.user.reset_password.invalid_link.app_error", nil, result.Err.Error())
+			return
+		} else {
+			recovery := result.Data.(*model.PasswordRecovery)
 
-		userId = data["user_id"]
+			if model.GetMillis()-recovery.CreateAt < model.PASSWORD_RECOVER_EXPIRY_TIME {
+				userId = recovery.UserId
+			} else {
+				c.Err = model.NewLocAppError("resetPassword", "api.user.reset_password.link_expired.app_error", nil, "")
+				return
+			}
 
-		timeStr = data["time"]
-		if len(timeStr) == 0 {
-			c.SetInvalidParam("resetPassword", "data:time")
-			return
+			go func() {
+				if result := <-Srv.Store.PasswordRecovery().Delete(userId); result.Err != nil {
+					l4g.Error("%v", result.Err)
+				}
+			}()
 		}
 	}
 
+	c.LogAuditWithUserId(userId, "attempt")
```

Comment:
Hmmm, it never was before we checked the hash but yeah it's probably a good idea to do it before